src/runtime/hexagon_remote/pipeline_context.h

```diff
@@ -5,49 +5,136 @@
 
     // Shared state
     pipeline_argv_t function;
     void **args;
     int result;
     bool running;
+    uint64_t last_did_work_time_us;
+    bool hvx_powered;
+
+    // Turn HVX power on if it isn't already on. Requires that the
+    // work_mutex is held.
+    int power_on_hvx_already_locked() {
+        if (hvx_powered) return 0;
+
+        HAP_power_request_t request;
+
+        request.type = HAP_power_set_apptype;
+        request.apptype = HAP_POWER_COMPUTE_CLIENT_CLASS;
+        int retval = HAP_power_set(NULL, &request);
+        if (0 != retval) {
+            log_printf("HAP_power_set(HAP_power_set_apptype) failed (%d)\n", retval);
+            return -1;
+        }
+
+        request.type = HAP_power_set_HVX;
+        request.hvx.power_up = TRUE;
+        retval = HAP_power_set(NULL, &request);
+        if (0 != retval) {
+            log_printf("HAP_power_set(HAP_power_set_HVX) failed (%d)\n", retval);
+            return -1;
+        }
+
+        request.type = HAP_power_set_mips_bw;
+        request.mips_bw.set_mips = TRUE;
+        request.mips_bw.mipsPerThread = 500;
+        request.mips_bw.mipsTotal = 1000;
+        request.mips_bw.set_bus_bw = TRUE;
+        request.mips_bw.bwBytePerSec = static_cast<uint64_t>(12000) * 1000000;
+        request.mips_bw.busbwUsagePercentage = 100;
+        request.mips_bw.set_latency = TRUE;
+        request.mips_bw.latency = 1;
+        retval = HAP_power_set(NULL, &request);
+        if (0 != retval) {
+            log_printf("HAP_power_set(HAP_power_set_mips_bw) failed (%d)\n", retval);
+            return -1;
+        }
+
+        hvx_powered = true;
+        return 0;
+    }
+
+    // Turn HVX power off if it's on. Requires that the work_mutex is
+    // held.
+    void power_off_hvx_already_locked() {
+        if (!hvx_powered) return;
+        HAP_power_request(0, 0, -1);
+        hvx_powered = false;
+    }
 
     void thread_main() {
         qurt_mutex_lock(&work_mutex);
         while (running) {
             qurt_cond_wait(&wakeup_thread, &work_mutex);
             if (function) {
+                power_on_hvx_already_locked();
                 result = function(args);
+                last_did_work_time_us = HAP_perf_get_time_us();
                 function = NULL;
                 qurt_cond_signal(&wakeup_caller);
             }
         }
         qurt_mutex_unlock(&work_mutex);
     }
 
+    void thread_watchdog() {
+        qurt_mutex_lock(&work_mutex);
+        // Every 100ms, wake up and check if HVX is powered on but no
+        // useful work has been done for 5 seconds. If so, turn off
+        // HVX.
+        const uint64_t poll_rate_us = 100 * 1000;
+        const uint64_t timeout = 5 * 1000 * 1000;
+        while (running) {
+            if (hvx_powered) {
+                uint64_t current_time = HAP_perf_get_time_us();
+                if (current_time > last_did_work_time_us + timeout) {
+                    power_off_hvx_already_locked();
+                }
```

Comment:
Should we try and shut down this thread while HVX is not powered? That way it doesn't cost any extra energy to run when HVX is already powered off.